### DISCUSSION TITLE
[I1] DUDA 2023-2-P1

### DISCUSSION THREAD

**Discussion by DaniloCordova13** [IIC2133-PUC/2024-2#127] [Preguntas Materia] [answered]
Hola! Estaba haciendo un ejercicio de una i pasada y hay algo que no me quedó del todo claro en una pauta de la pregunta 1, específicamente de la parte d.
![IMG_1858](https://github.com/user-attachments/assets/5326e5da-de88-4b7b-8b22-0b48b5a447f9)
Con respecto a la complejidad del peor caso, mi análisis fue que en cada iteración del hule (en total son n) se hace merge, pero cada merge no tiene complejidad n exacta, ya que se van agregando los valores de a poco
![IMG_1860](https://github.com/user-attachments/assets/801f5b96-fdb1-4e80-8aba-25b7b94b7e69)
Al ser constantes, salen de la complejidad quedando igual en O(n) (solo en merge), o es esa complejidad porque en la iteración final se recorren todos?. Al final puse que la complejidad es O(n^2) por lo de las constantes, pero no me quedó muy claro si es válido ese razonamiento
![IMG_1861](https://github.com/user-attachments/assets/6c9d05be-a86c-4a81-9420-2b3520ccf54d)
En la pauta justifica a través del foreach, pero no me convenció del todo. De hecho para ser honesto no entendí porque eso hace que la complejidad sea O(n^2). 
Gracias de antemano ^^

**Comment by Segismundios** [comment, marked as answer]
Hola @DaniloCordova13. Para entender mejor la complejidad del algoritmo, considera lo siguiente:

En el peor de los casos, el while se ejecuta tantas veces como elementos tenga inicialmente la lista A. Es decir, si A comienza con n elementos, el while iterará n veces, eliminando un elemento de A en cada iteración.

Por otro lado, como el tamaño de A disminuye en 1 con cada iteración del while, el foreach  recorrerá primero n elementos, luego n-1, después n-2, y así sucesivamente hasta que solo quede un elemento en A.

Entonces, el número total de operaciones realizadas por el foreach a lo largo de todas las iteraciones del while será:  $n + n-1 + \dots  + 1$.

Esta suma se puede expresar matemáticamente como:
$$\sum_{i=1}^{n} i = \frac{n \cdot (n+1)}{2} = \frac{n^2+n}{2}$$
​
Por lo tanto, la complejidad en el peor caso del foreach es $O(n^2)$.

También creo que es importante mencionar que en cada iteración se agrega solo 1 elemento a la lista B. Esto significa que:

El primer merge se realiza con 1 elemento en total.
Luego, se hace un merge con 2 elementos (ahora C tiene 1 elemento y B tiene 1 elemento)
Y así sucesivamente, hasta que el último merge se realiza con $n$ elementos (ahora C tiene n-1 elementos y B tiene 1 elemento).
Entonces, la complejidad total de todos los merge será: $O(1 + 2 + \dots + n)$, lo que equivale a $O(\sum_{i=1}^{n}i)$.
Como vimos antes, esto es igual a $O(n^2)$.

Por lo tanto, podemos concluir que la complejidad total del algoritmo, considerando tanto el foreach como los merge, será $O(n^2 + n^2) = O(2n^2) = O(n^2)$ en el peor caso.

Saludos!
